hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -323,44 +325,44 @@ protected final int calculatePartitionHash(Object obj, PartitioningStrategy stra
     }
 
     protected final boolean safeRegister(final Class type, final Serializer serializer) {
-        return safeRegister(type, createSerializerAdapter(serializer, this));
+        return safeRegister(type, asStreamSerializer(serializer));
     }
 
-    protected final boolean safeRegister(final Class type, final SerializerAdapter serializer) {
+    protected final boolean safeRegister(final Class type, final StreamSerializer serializer) {
         if (constantTypesMap.containsKey(type)) {
             throw new IllegalArgumentException("[" + type + "] serializer cannot be overridden!");
         }
-        SerializerAdapter current = typeMap.putIfAbsent(type, serializer);
-        if (current != null && current.getImpl().getClass() != serializer.getImpl().getClass()) {
+        StreamSerializer current = typeMap.putIfAbsent(type, serializer);
+        if (current != null && current.getClass() != serializer.getClass()) {
```

Comment:
Yes. By the way, there is no such test in system. I just wrote a new test and shared with you to give an idea. You may want to add this test in your pr.